### PR TITLE
Enable config driven boto timeouts

### PR DESCRIPTION
**Pull request by anantsimran**
To be tackled in further PRs:
1. Unit test cases
2. More robust code. i.e. it would fail now if default values are not provided
3. Change file download logs to debug

### CODE REVIEW

File: tests/.wickerconfig.test.json [hunk 1]
```diff
@@ -1,10 +1,19 @@
 {
     "aws_s3_config": {
       "s3_datasets_path": "s3://SOME_FAKE_PATH",
-      "region": "us-west-2"
+      "region": "us-west-2",
+      "boto_config": {
+        "max_pool_connections":10,
+        "read_timeout": 140,
```
Comment (aalavian) [inline tests/.wickerconfig.test.json:7]: nit: specify unit here and elsewhere: `read_timeout_s`

File: tests/.wickerconfig.test.json [hunk 2]
```diff
@@ -1,10 +1,19 @@
 {
     "aws_s3_config": {
       "s3_datasets_path": "s3://SOME_FAKE_PATH",
-      "region": "us-west-2"
+      "region": "us-west-2",
+      "boto_config": {
+        "max_pool_connections":10,
```
Comment (aalavian) [inline tests/.wickerconfig.test.json:6]: nit: `: 10,`

File: wicker/core/filelock.py [hunk 3]
```diff
@@ -4,8 +4,11 @@
 import fcntl
 import os
 import signal
+import logging
```
Comment (aalavian) [inline wicker/core/filelock.py]: hmm, remove all debug logs from this file

File: wicker/core/storage.py [hunk 4]
```diff
@@ -34,8 +37,15 @@ def __init__(self, session: Optional[boto3.session.Session] = None) -> None:
         unit tests to easily mock / patch the S3 client or to utilize the S3 Stubber. Unit tests
         might also find it convenient to mock or patch member functions on instances of this class.
         """
+        boto_config = get_config().aws_s3_config.boto_config
+        boto_client_config= botocore.config.Config(
```
Comment (aalavian) [inline wicker/core/storage.py]: nit: `boto_client_config = `

File: wicker/core/filelock.py [hunk 5]
```diff
@@ -28,6 +31,7 @@ def timeout_handler(signum: Any, frame: Any) -> None:
             """Handles receiving a SIGALRM by throwing a TimeoutError with a useful
             error message
             """
+            logging.error("Found the timeout error")
```
Comment (aalavian) [inline wicker/core/filelock.py:34]: `logging.error(f"Found the timeout error for file {lockfile}")`

File: wicker/core/storage.py [hunk 6]
```diff
@@ -34,8 +37,15 @@ def __init__(self, session: Optional[boto3.session.Session] = None) -> None:
...
+            max_pool_connections=boto_config.max_pool_connections,
+            read_timeout=boto_config.read_timeout,
+            connect_timeout=boto_config.connect_timeout,
+        )
         self.session = boto3.session.Session() if session is None else session
-        self.client = self.session.client("s3")
+        self.client = self.session.client("s3", config=boto_client_config)
+        self.read_timeout= get_config().s3_storage_config.timeout
```
Comment (aalavian) [inline wicker/core/storage.py]: nit: `self.read_timeout = ...`

File: wicker/core/storage.py [hunk 7]
```diff
@@ -72,6 +82,18 @@ def check_exists_s3(self, input_path: str) -> bool:
         except ClientError:
             return False
 
+    @retry(Exception, tries=get_config().s3_storage_config.retries, backoff=5, delay=4, logger=logging)
```
Comment (aalavian) [inline wicker/core/storage.py]: Could we have backoff and delay also in configs?

File: wicker/core/storage.py [hunk 8]
```diff
@@ -72,6 +82,18 @@ def check_exists_s3(self, input_path: str) -> bool:
...
+    def download_with_retries(self, bucket:str, key: str, local_path: str,s3_input_path: str):
+        try:
+            with time_limit(self.read_timeout):
+                logging.info(f"Trying to download {bucket} {key}")
```
Comment (aalavian) [inline wicker/core/storage.py]: Let's remove debug log like this one and the one below.

File: wicker/core/utils.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import contextlib
+import signal
+
+@contextlib.contextmanager
+def time_limit(seconds):
+    def signal_handler(signum, frame):
+        raise TimeoutError("Timed out!")
```
Comment (aalavian) [inline wicker/core/utils.py:8]: Let's have a more informative message: maybe "Timing out in trying to download file in Wicker"

File: wicker/core/storage.py [hunk 10]
```diff
@@ -72,6 +82,18 @@ def check_exists_s3(self, input_path: str) -> bool:
...
 
+    @retry(Exception, tries=get_config().s3_storage_config.retries, backoff=5, delay=4, logger=logging)
+    def download_with_retries(self, bucket:str, key: str, local_path: str,s3_input_path: str):
+        try:
+            with time_limit(self.read_timeout):
+                logging.info(f"Trying to download {bucket} {key}")             
+                self.client.download_file(bucket, key, local_path)
+                logging.info(f"Downloaded {s3_input_path} on to {local_path}")
+        except Exception as e:
+            logging.error(f'Failed to download s3 path:{s3_input_path}')
```
Comment (aalavian) [inline wicker/core/storage.py]: nit: `path: {s3_...`

File: wicker/core/storage.py [hunk 11]
```diff
@@ -99,10 +121,10 @@ def fetch_file_s3(self, input_path: str, local_prefix: str, timeout_seconds: int
                     # Long term, we would also add a size check or md5sum comparison against the object in S3.
                     filedir = os.path.split(local_path)[0]
                     os.makedirs(filedir, exist_ok=True)
-                    self.client.download_file(bucket, key, local_path)
-
+                    self.download_with_retries(bucket=bucket, key=key, local_path=local_path, s3_input_path=input_path)
                     with open(success_marker, "w"):
                         pass
+            logging.debug("Exited the file lock successfully")
```
Comment (aalavian) [inline wicker/core/storage.py:126]: Remove debug log.

File: wicker/core/storage.py [hunk 12]
```diff
@@ -72,6 +82,17 @@ def check_exists_s3(self, input_path: str) -> bool:
         except ClientError:
             return False
 
+    @retry(Exception, tries=get_config().s3_storage_config.retries, backoff=5, delay=4, logger=logging)
+    def download_with_retries(self, bucket: str, key: str, local_path: str, s3_input_path: str):
```
Comment (aalavian) [inline wicker/core/storage.py:86]: Let's drop `s3_input_path` and use `key` instead.

File: tests/.wickerconfig.test.json [hunk 13]
```diff
@@ -1,10 +1,19 @@
...
+        "read_timeout": 140,
+        "connect_timeout": 140
+      }
     },
     "dynamodb_config": {
       "table_name": "fake-table-name",
       "region": "us-west-2"
+    },
+    "s3_storage_config":{
```
Comment (aalavian) [inline tests/.wickerconfig.test.json:15]: It looks like this is generalizable to any client, let's rename to `storage_download_config`?

File: wicker/core/storage.py [hunk 14]
```diff
@@ -72,6 +82,17 @@ def check_exists_s3(self, input_path: str) -> bool:
...
+        try:
+            with time_limit(self.read_timeout):
+                logging.info(f"Trying to download {bucket} {key}")
```
Comment (aalavian) [inline wicker/core/storage.py:89]: Let's drop this info log and the one on line 91.

File: wicker/core/storage.py [hunk 15]
```diff
@@ -72,6 +82,17 @@ def check_exists_s3(self, input_path: str) -> bool:
         except ClientError:
             return False
 
+    @retry(Exception, tries=get_config().s3_storage_config.retries, backoff=5, delay=4, logger=logger)
```
Comment (aalavian) [inline wicker/core/storage.py:85]: Could we also have backoff and delay in configs too?

**Comment by anantsimran** [comment]
meow @aalavian @isaak-willett .
Tested locally and on prod here: https://us-west-2.console.aws.amazon.com/sagemaker/home?region=us-west-2#/jobs/anantsingh-2023-09-22-0007

File: wicker/core/filelock.py [hunk 16]
```diff
@@ -38,11 +42,13 @@ def timeout_handler(signum: Any, frame: Any) -> None:
             # Raise a SIGALRM in `seconds` number of seconds, and then yield to code
             # in the context manager's code-block.
             signal.alarm(seconds)
+            logging.debug("Was not able to find an error on the file yet")
```
Comment (aalavian) [inline wicker/core/filelock.py:45]: Remove debug log here and elsewhere in this file.

File: wicker/core/config.py [hunk 17]
```diff
@@ -23,33 +23,65 @@ def from_json(cls, data: Dict[str, Any]) -> WickerWandBConfig:
...
+    @classmethod
+    def from_json(cls, data: Dict[str, Any]) -> S3StorageConfig:
+        return cls(
+            retries=data["retries"],
+            timeout=data["timeout"],
+        )
+
+
 @dataclasses.dataclass(frozen=True)
 class WickerConfig:
```
Comment (isaak-willett) [inline wicker/core/config.py:72]: Do we need to add the boto config to this object too? I think we should probably have it

File: wicker/core/filelock.py (hunk 5 above)
Comment (isaak-willett) [inline wicker/core/filelock.py:34]: Remove

File: wicker/core/filelock.py (hunk 16 above)
Comment (isaak-willett) [inline wicker/core/filelock.py:45]: Remove

File: wicker/core/filelock.py [hunk 18]
```diff
@@ -38,11 +42,13 @@ def timeout_handler(signum: Any, frame: Any) -> None:
...
             yield
         # Make sure to always restore the original SIGALRM handler
         finally:
             signal.alarm(0)
             signal.signal(signal.SIGALRM, original_handler)
+            logging.debug("Exiting the timeout handler with successful messaging")
```
Comment (isaak-willett) [inline wicker/core/filelock.py:51]: Remove

File: wicker/core/filelock.py [hunk 19]
```diff
@@ -2,10 +2,13 @@
 
 import contextlib
 import fcntl
+import logging
 import os
 import signal
 from typing import Any, Iterator, Optional
 
+logger = logging.getLogger(__name__)
+
```
Comment (isaak-willett) [inline wicker/core/filelock.py:11]: Remove

File: wicker/core/filelock.py [hunk 20]
```diff
@@ -2,10 +2,13 @@
 
 import contextlib
 import fcntl
+import logging
```
Comment (isaak-willett) [inline wicker/core/filelock.py:5]: Remove

File: wicker/core/filelock.py [hunk 21]
```diff
@@ -72,11 +78,15 @@ def __enter__(self) -> SimpleUnixFileLock:
         fd = os.open(self._lock_file, open_mode)
         try:
             with _timeout(self.timeout_seconds, self._lock_file):
+                logging.debug("Inside the locking mechanism")
```
Comment (isaak-willett) [inline wicker/core/filelock.py:81]: Remove

File: wicker/core/filelock.py [hunk 22]
```diff
@@ -72,11 +78,15 @@ def __enter__(self) -> SimpleUnixFileLock:
...
                 fcntl.lockf(fd, fcntl.LOCK_EX)
-        except (IOError, OSError, TimeoutError):
+                logging.debug("Acquired the lock on the file, waiting for s3 to finish download")
+        except (IOError, OSError, TimeoutError) as error:
+            logging.debug(f"Encountered error within lock {error}")
             os.close(fd)
             raise
         else:
+            logging.debug("Releasing lock as the enter clause has ended")
```
Comment (isaak-willett) [inline wicker/core/filelock.py:89]: Remove

**Review by aalavian — COMMENTED** [review]
This is great, thanks @anantsimran for introducing this.

File: wicker/core/filelock.py [hunk 23]
```diff
@@ -72,11 +78,15 @@ def __enter__(self) -> SimpleUnixFileLock:
         fd = os.open(self._lock_file, open_mode)
         try:
             with _timeout(self.timeout_seconds, self._lock_file):
+                logging.debug("Inside the locking mechanism")
                 fcntl.lockf(fd, fcntl.LOCK_EX)
-        except (IOError, OSError, TimeoutError):
+                logging.debug("Acquired the lock on the file, waiting for s3 to finish download")
+        except (IOError, OSError, TimeoutError) as error:
+            logging.debug(f"Encountered error within lock {error}")
```
Comment (isaak-willett) [inline wicker/core/filelock.py:85]: Remove changes here

File: wicker/core/config.py (hunk 17 above)
Comment (isaak-willett) [reply wicker/core/config.py:72]: Sorry missed it in the config above, ignore this comment, you put it in a good place

File: wicker/core/storage.py [hunk 24]
```diff
@@ -72,6 +82,17 @@ def check_exists_s3(self, input_path: str) -> bool:
...
+        try:
+            with time_limit(self.read_timeout):
+                logging.info(f"Trying to download {bucket} {key}")
+                self.client.download_file(bucket, key, local_path)
+                logging.info(f"Downloaded {s3_input_path} on to {local_path}")
+        except Exception as e:
+            logging.error(f"Failed to download s3 path:{s3_input_path}")
+            raise e
+
     def fetch_file_s3(self, input_path: str, local_prefix: str, timeout_seconds: int = 120) -> str:
```
Comment (isaak-willett) [inline wicker/core/storage.py:96]: Do you want to add an interface for timeout seconds too or are you comfortable with this? Just curious no need to implement

**Review by isaak-willett — APPROVED** [review]
Looks good, just git rid of the logging and the Alborz comments

**Comment by isaak-willett** [comment]
This is great thank you for doing this!

File: wicker/core/filelock.py (hunk 20 above)
Comment (anantsimran) [reply wicker/core/filelock.py:5]: Done.

File: wicker/core/storage.py (hunk 12 above)
Comment (anantsimran) [reply wicker/core/storage.py:92]: Makes sense. Done.

File: wicker/core/storage.py (hunk 24 above)
Comment (anantsimran) [reply wicker/core/storage.py:102]: I am not sure I get you here. What would an interface would timeout seconds look like?

File: tests/.wickerconfig.test.json (hunk 1 above)
Comment (anantsimran) [reply tests/.wickerconfig.test.json]: Lint issue. Corrected.

File: wicker/core/storage.py (hunk 7 above)
Comment (anantsimran) [reply wicker/core/storage.py]: Done.

File: wicker/core/utils.py (hunk 9 above)
Comment (anantsimran) [reply wicker/core/utils.py]: Done.

File: wicker/core/storage.py (hunk 15 above)
Comment (anantsimran) [reply wicker/core/storage.py]: Done.

**Pull request merged**